research/picai/fl_nnunet/nnunet_client.py

```diff
@@ -0,0 +1,607 @@
+        return self.loss(pred, target)
+
+
+class nnUNetClient(BasicClient):
+    def __init__(
+        self,
+        dataset_id: int,
+        data_identifier: Optional[str],
+        plans_identifier: Optional[str],
+        always_preprocess: bool = True,
+        **kwargs: Any,
+    ) -> None:
+        """
+        _summary_
+
+        Args:
+            dataset_id (int): _description_
+            data_identifier (Optional[str]): _description_
+            plans_identifier (Optional[str]): _description_
+            always_preprocess (bool, optional): _description_. Defaults to True.
+        """
+        super().__init__(**kwargs)
+        self.dataset_id: int = dataset_id
+        self.dataset_name = convert_id_to_dataset_name(self.dataset_id)
+        self.dataset_json = load_json(join(nnUNet_raw, self.dataset_name, "dataset.json"))
+        self.data_identifier = data_identifier
+        self.device = kwargs["device"]
+        self.always_preprocess: bool = always_preprocess
+        self.plans_name = plans_identifier
+
+        self.nnunet_trainer: nnUNetTrainer
+
+    def get_data_loaders(self, config: Config) -> Tuple[DataLoader, DataLoader]:
+        """
+        Gets the nnunet dataloaders and wraps them in another class to make them
+        pytorch iterators
+
+        Args:
+            config (Config): The config file from the server
+
+        Returns:
+            Tuple[DataLoader, DataLoader]: A tuple of length two. The client
+                train and validation dataloaders as pytorch dataloaders
+        """
+        # flwr 1.9.0 now raises an error any time a process ends.
+        # To prevent errors due to this since dataloaders use multiprocessing
+        # lets ovverride that behaviour before the child processes are created
+        fl_sigint_handler = signal.getsignal(signal.SIGINT)
+        fl_sigterm_handler = signal.getsignal(signal.SIGTERM)
+        signal.signal(signal.SIGINT, og_sigint_handler)
+        signal.signal(signal.SIGTERM, og_sigterm_handler)
+
+        # Get the nnunet dataloader iterators
+        with nostdout():
+            train_loader, val_loader = self.nnunet_trainer.get_dataloaders()
+
+        # Set the signal handlers back to what they were for flwr
+        signal.signal(signal.SIGINT, fl_sigint_handler)
+        signal.signal(signal.SIGTERM, fl_sigterm_handler)
+
+        # The batchgenerators package used under the hood by the dataloaders
+        # creates an additional stream handler for the root logger
+        # Therefore all logs get printed twice, We can fix this by clearing the
+        # root logger handlers.
+        root_logger = logging.getLogger()
+        root_logger.handlers.clear()
+
+        # By default nnunet dataloaders are infinite, so if we are training by
+        # num epochs we need to make them not infinite anymore
+        if config["local_epochs"] is not None:
+            infinite = False
```

Comment:
I believe the nnunet dataloaders are infinite by default. But if we are using train_by_epochs, we need to set infinite to false so that the epoch doesn't last forever. Technically when infinite is set to False, we are still using an infinite dataloader that randomly samples the dataset, we just generate a stop iteration after {num_samples/batch_size} steps to simulate the completion of an epoch

Reply:
Although there actually is a bug here, the infinite variable is not set in the case where we are training by sets. I need to add an else: infinite=True